### PR TITLE
Fix wrapper for psycopg2 connection so type check does not fail

### LINKED ISSUE #44: psycopg2 throws an exception in sqlalchemy when updating to 2.4.0
After the version in our project was bumped from 2.3.0 to 2.4.0 we started getting the following error for all methods that were touching the orm layer
```
TypeError: argument 2 must be a connection, cursor or None
....
  File "sqlalchemy/orm/query.py", line 2469, in first
    ret = list(self[0:1])
  File "sqlalchemy/orm/query.py", line 2292, in __getitem__
    return list(res)
  File "flask_sqlalchemy_cache/core.py", line 50, in __iter__
    return BaseQuery.__iter__(self)
  File "sqlalchemy/orm/query.py", line 2571, in __iter__
    return self._execute_and_instances(context)
  File "sqlalchemy/orm/query.py", line 2584, in _execute_and_instances
    close_with_result=True)
  File "sqlalchemy/orm/query.py", line 2575, in _connection_from_session
    **kw)
  File "sqlalchemy/orm/session.py", line 893, in connection
    execution_options=execution_options)
  File "sqlalchemy/orm/session.py", line 898, in _connection_for_bind
    engine, execution_options)
  File "sqlalchemy/orm/session.py", line 334, in _connection_for_bind
    conn = bind.contextual_connect()
  File "sqlalchemy/engine/base.py", line 2039, in contextual_connect
    self._wrap_pool_connect(self.pool.connect, None),
  File "sqlalchemy/engine/base.py", line 2074, in _wrap_pool_connect
    return fn()
  File "sqlalchemy/pool.py", line 376, in connect
    return _ConnectionFairy._checkout(self)
  File "sqlalchemy/pool.py", line 713, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "sqlalchemy/pool.py", line 480, in checkout
    rec = pool._do_get()
  File "sqlalchemy/pool.py", line 1060, in _do_get
    self._dec_overflow()
  File "sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "sqlalchemy/pool.py", line 1057, in _do_get
    return self._create_connection()
  File "sqlalchemy/pool.py", line 323, in _create_connection
    return _ConnectionRecord(self)
  File "sqlalchemy/pool.py", line 455, in __init__
    pool.dispatch.connect(self.connection, self)
  File "sqlalchemy/event/attr.py", line 256, in __call__
    fn(*args, **kw)
  File "sqlalchemy/engine/strategies.py", line 156, in on_connect
    do_on_connect(conn)
  File "sqlalchemy/dialects/postgresql/psycopg2.py", line 668, in on_connect
    fn(conn)
  File "sqlalchemy/dialects/postgresql/psycopg2.py", line 632, in on_connect
    extras.register_uuid(None, conn)
  File "psycopg2/extras.py", line 487, in register_uuid
    _ext.register_type(_ext.UUID, conn_or_curs)

```

I guess after the client hooks that were added to in this change https://github.com/uber-common/opentracing-python-instrumentation/commit/fa35a88aa27501656455fb57fef3a3911273c561 it broke a contract somewhere in sqlalchemy/psycopg2

### PR DESCRIPTION
**Pull request by isaachier**
Fixes #44.

### CODE REVIEW

File: opentracing_instrumentation/client_hooks/_dbapi2.py [hunk 1]
```diff
@@ -161,6 +161,10 @@ def __init__(self, connection, module_name, connect_params):
             connect_params=connect_params
         )
 
+        # Tip suggested here:
+        # https://gist.github.com/mjallday/3d4c92e7e6805af1e024.
+        self._sqla_unwrap = connection
```
Comment (yurishkuro) [inline opentracing_instrumentation/client_hooks/_dbapi2.py]: without a test, impossible to see the impact. Can we reproduce the problem reported in #44 in a test?

Comment (isaachier) [reply opentracing_instrumentation/client_hooks/_dbapi2.py]: Not without sqlachemy connecting to a postgres database using psycopg2.

**Comment by yurishkuro** [comment]
we can run postgres in Travis, same was as we run Redis - https://github.com/uber-common/opentracing-python-instrumentation/blob/master/.travis.yml#L15

**Comment by isaachier** [comment]
K let me try that

File: tests/opentracing_instrumentation/test_postgres.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+class User(object):
+
+    def __init__(self, name, fullname, password):
+        self.name = name
+        self.fullname = fullname
+        self.password = password
+
+
+mapper(User, user)
```
Comment (yurishkuro) [inline tests/opentracing_instrumentation/test_postgres.py]: please move up next to the other vars

File: opentracing_instrumentation/client_hooks/_dbapi2.py [hunk 3]
```diff
@@ -161,13 +161,19 @@ def __init__(self, connection, module_name, connect_params):
...
+
     def __enter__(self):
         with func_span('%s:begin_transaction' % self._module_name):
-            cursor = self.__wrapped__.__enter__()
+            connection = self.__wrapped__.__enter__()
 
-        return CursorWrapper(cursor=cursor,
-                             module_name=self._module_name,
-                             connect_params=self._connect_params)
+        return ConnectionWrapper(connection=connection,
```
Comment (yurishkuro) [inline opentracing_instrumentation/client_hooks/_dbapi2.py]: can you elaborate on this change?

File: tests/opentracing_instrumentation/test_postgres.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    Session.configure(bind=engine)
+    try:
+        yield Session()
+    except:
+        pass
+
+
+@pytest.fixture(autouse=True)
+def patch_postgres():
+    psycopg2.install_patches()
```
Comment (yurishkuro) [inline tests/opentracing_instrumentation/test_postgres.py]: this probably works now because you only have one test, but in theory a fixture should undo the changes.

Comment (yurishkuro) [reply tests/opentracing_instrumentation/test_postgres.py]: Doesn't have to be in this PR, but would be nice to extend this test to MySQL.

File: .travis.yml [hunk 5]
```diff
@@ -6,17 +6,22 @@ matrix:
...
 
-before_install:
-  - bash travis/install-redis.sh
-
 install:
   - make install-ci
 
+services:
+  - postgresql
+  - redis
```
Comment (yurishkuro) [inline .travis.yml]: 🎉

File: tests/opentracing_instrumentation/test_postgres.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    def __init__(self, name, fullname, password):
+        self.name = name
+        self.fullname = fullname
+        self.password = password
+
+
+mapper(User, user)
+
+def test_db(tracer, engine, session):
```
Comment (yurishkuro) [inline tests/opentracing_instrumentation/test_postgres.py]: can be another PR, but is there a way to skip this test if POSTGRES not running?

File: tests/opentracing_instrumentation/test_postgres.py (hunk 2 above)
Comment (isaachier) [reply tests/opentracing_instrumentation/test_postgres.py]: OK

File: opentracing_instrumentation/client_hooks/_dbapi2.py (hunk 3 above)
Comment (isaachier) [reply opentracing_instrumentation/client_hooks/_dbapi2.py]: Weird. Meant to revert this. Strictly speaking, dbapi2 is not exactly what we are modeling here. Note the example given in the comment above (https://github.com/uber-common/opentracing-python-instrumentation/pull/55/files/d7f5d4d0c0892e98b7a256d0b7933f5814c661de#diff-2b264c7b6b41ea2c077d7e1125a983b2R155):

```
         with MySQLdb.connect(...) as cursor:
             cursor.execute(...)
```

I believe dbapi2 specifies this:

```
         with MySQLdb.connect(...) as conn:
             with conn.cursor() as cursor:
               cursor.execute(...)
```

I was going to make these changes, but SQLAlchemy seems to accept both and I didn't want to break backwards compatibility.

File: tests/opentracing_instrumentation/test_postgres.py (hunk 4 above)
Comment (isaachier) [reply tests/opentracing_instrumentation/test_postgres.py]: OK I can extend to MySQL, but you won't get the same failing case. I can reproduce the error locally when I remove the `__getattr__` method from connection.

Comment (isaachier) [reply tests/opentracing_instrumentation/test_postgres.py]: About the fixture, I agree, but no undo was currently in the postgres module. Not sure if that was intentional.

File: tests/opentracing_instrumentation/test_postgres.py (hunk 6 above)
Comment (isaachier) [reply tests/opentracing_instrumentation/test_postgres.py]: Probably a good idea. What about redis?

File: tests/opentracing_instrumentation/test_postgres.py (hunk 2 above)
Comment (isaachier) [reply tests/opentracing_instrumentation/test_postgres.py]: Actually I can't move it because the class is only defined immediately before this. I blame the SQLAlchemy API for the mess of variables in module scope.

**Comment by yurishkuro** [comment]
lgtm to merge - cc @black-adder

File: tests/opentracing_instrumentation/test_postgres.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    opentracing.tracer = t
+
+    try:
+        yield t
+    except:
+        opentracing.tracer = old_tracer
+
+
+def check_span():
+    print([span.__dict__ for span in opentracing.tracer.recorder.get_spans()])
```
Comment (black-adder) [inline tests/opentracing_instrumentation/test_postgres.py]: are these prints for debug?

File: .travis.yml [hunk 8]
```diff
@@ -6,17 +6,22 @@ matrix:
   include:
   - python: '2.7'
     env: COVER=1
-  - python: '3.3'
+  # For some reason not working.
+  # - python: '3.3'
```
Comment (black-adder) [inline .travis.yml]: also, what was up with this? Was it transient?

File: tests/opentracing_instrumentation/test_postgres.py (hunk 7 above)
Comment (isaachier) [reply tests/opentracing_instrumentation/test_postgres.py]: Good catch!

File: .travis.yml (hunk 8 above)
Comment (isaachier) [reply .travis.yml]: not temporary as far as i know. hold on.

Comment (isaachier) [reply .travis.yml]: From Travis log:

```
pip install -r requirements.txt
Obtaining file:///home/travis/build/uber-common/opentracing-python-instrumentation (from -r requirements.txt (line 3))
Exception:
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/commands/install.py", line 310, in run
    wb.build(autobuilding=True)
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/wheel.py", line 750, in build
    self.requirement_set.prepare_files(self.finder)
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/req/req_set.py", line 370, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/req/req_set.py", line 489, in _prepare_file
    abstract_dist.prep_for_dist()
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/req/req_set.py", line 127, in prep_for_dist
    self.req_to_install.run_egg_info()
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/req/req_install.py", line 402, in run_egg_info
    self.setup_py, self.link,
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pip/req/req_install.py", line 372, in setup_py
    import setuptools  # noqa
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/setuptools/__init__.py", line 12, in <module>
    import setuptools.version
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/setuptools/version.py", line 1, in <module>
    import pkg_resources
  File "/home/travis/virtualenv/python3.3.6/lib/python3.3/site-packages/pkg_resources/__init__.py", line 85, in <module>
    raise RuntimeError("Python 3.4 or later is required")
RuntimeError: Python 3.4 or later is required
make: *** [install-deps] Error 2
```

**Comment by isaachier** [comment]
I still have to deploy this in production to test.
